### ISSUE TITLE
💬 RFC: Add knip for CI

### ISSUE THREAD

**Issue by drodil** [backstage/backstage#21948]
### 🔖 Need

There are currently a lot of issues with the packages (unused/duplicate exports, unused packages etc.). 

Here is [knip](https://knip.dev/) output for the somewhat current backstage repository. Some might be false positives but they should be checked and the tool configured properly.

[knip_findings.txt](https://github.com/backstage/backstage/files/13724292/knip_findings.txt)

### 🎉 Proposal

I suggest adding [knip](https://knip.dev/) tool to be part of the CI/CD pipeline of backstage. 

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by awanlin** [comment]
Hi @drodil, this seems like a really good idea. Something I always thought about with my Backstage instances but never had the time to really look at. Over time you add packages but often I know I don't always remember to removed them.

I'm looking over the report you shared and their is a lot of noise, any chance you can clean it up so we get a closer example of what the reality is? Some items that for sure can be filtered out:

- The `/migrations` folder - these for sure are being used but probably not in a way that can be easily identified
- All the `*.stories.tsx` files - these are for StoryBook and are being used, hopefully :)
- Exclude the `/contrib` folder as these are contributed documentation
- The following files all are being used in some way:
   - `config.d.ts`
   - `knexfile.js`
   - `setupTests.ts`
   - `run.ts`
   - `standaloneServer.ts`
   - `standaloneApplication.ts`

**Comment by freben** [comment]
The amount of false positives in there is off the hook though.

- 100 of them were `/dev/` folder things (these are used)
- 194 were `setupTests.ts` (these are used)
- 69 were `.stories.*` Storybook files (these are used)
- 2186 were `eslintrc` related (doesn't apply)
- 52 were `run.ts`/`standaloneServer.ts` related (these are used)
- 60 were migrations (these are used)
- 85 were `config.d.ts` (these are used)
- 50 are `core-app-api` deps used by the dev endpoint
- etc etc

**Comment by drodil** [comment]
Yeah, there are false positives since I just ran the tool without any configuration. When doing PR for this, there should be proper config in place. In case this makes sense of course :)

**Comment by freben** [comment]
Would you envision this to be run in CI and fail builds if it isn't clean? I would presume that it'd need to be introduced on a package-by-package basis then because the amount of work seems to be pretty significant

**Comment by drodil** [comment]
Well you could start by allowing, let's say 300 violations with the max issues parameter: https://knip.dev/reference/cli#--max-issues

And then make it go down by the time. I think it's something that can't and should not be fixed at one go, but maybe to reduce the amount by time and check that there's no new issues coming in

**Comment by awanlin** [comment]
Could you create an action that would update an issue comment on a regular schedule with the report data? So we would have a Knip Tracking issue and then the action would update the same comment with the latest report data. There's an action to do something like that: https://github.com/marketplace/actions/create-or-update-comment. That would allow us to track this and get it down to a state where it could be part of the CI.

**Comment by drodil** [comment]
Yeah, that could be another "Help wanted" issue where the updates would go to. And people fixing the issues could then link their PRs to that issue. 

Anyway, the tool is new for me also so some studying is needed on how to configure it and if it's possible to have folder-specific configurations. I did very simple config to the root and it already cut down number of findings quite a lot:

```json
{
  "ignore": ["**/dev/**", "setupTests.ts", ".*stories.*", ".eslintrc.js", "run.ts", "standaloneServer.ts", "**/migrations/**", "config.d.ts", "**/core-app-api/**"]
}

```

**Comment by Rugvip** [comment]
Overall a big 👍 to introducing this. In particular I do already see some troublesome extraneous dependencies, would love to have that more properly locked down.

It would be important that we have some sort of allow list though, apart for some global filters I think we'd also want to have so package-specific filters to allow certain deps, for example the DB drivers like `pg`. Given that there's support for dynamic configuration we can prolly do more or less what we want there though so even if knip doesn't support something out of the box it's likely we'd be able to build something to fit our needs.

**Comment by freben** [comment]
For preparation, I got inspired to do https://github.com/backstage/backstage/pull/21958

**Comment by drodil** [comment]
Made the initial version of the config and cron job in #21963 

Work in progress and need your input for that so comments welcome for that PR

**Comment by drodil** [comment]
@freben maybe re-open this and discuss:

- [ ] How to add the actual CI part to update the reports
- [ ] If more checks should be added apart from dependencies and unlisted dependencies (https://knip.dev/reference/cli#--include)
